src/main/ruby/truffleruby/core/posix.rb

```diff
@@ -425,11 +425,62 @@ def self.read_string_native(io, length)
     end
   end
 
-  def self.read_to_buffer_native(io, length)
+  def self.execute_posix_read(io, buffer, length)
+    fd = io.fileno
+    return Truffle::POSIX.read(fd, buffer, length) unless io.timeout
+
+    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + io.timeout
+
+    loop do
+      current_timeout = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      raise IO::TimeoutError if current_timeout < 0
```

Comment:
This feels wrong because then we might not even try once to call poll()

Reply:
And indeed it should try at least once:
```
irb(main):001:0> r,w=IO.pipe
=> [#<IO:fd 5>, #<IO:fd 6>]
irb(main):002:0> w.timeout = 0.001
=> 0.001
irb(main):003:0> w.write "a"
=> 1
irb(main):004:0> w.timeout = 0
=> 0
irb(main):005:0> w.write "a"
=> 1
```

Reply:
Yes, updated.